src/components/RichTextEditor/RichTextEditor.tsx

```diff
@@ -0,0 +1,292 @@
+
+    const {
+        styles,
+        attributes,
+        update: popperUpdate,
+    } = usePopper(selectionRectRef.current, inlineToolbarRef.current, {
+        placement: "top",
+        modifiers: [
+            { name: "arrow", options: { element: inlineToolbarArrowRef.current, padding: 10 } },
+            {
+                name: "offset",
+                options: {
+                    offset: [0, 8],
+                },
+            },
+            {
+                name: "flip",
+                options: {
+                    fallbackPlacements: ["bottom", "right"],
+                },
+            },
+        ],
+    });
+
+    useClickOutside(inlineToolbarRef.current, () => {
+        if (showInlineToolbar) {
+            setSavedSelection(null);
+
+            if (editorArea !== RichTextEditorArea.Normal) {
+                setEditorArea(RichTextEditorArea.Normal);
+            }
+        }
+    });
+
+    const onEditorChange = async (value: EditorState): Promise<void> => {
+        const currentContent = editorState.getCurrentContent();
+        const newContent = value.getCurrentContent();
+
+        // New character added to the editor
+        if (currentContent !== newContent) {
+            emitContentChanged();
+        }
+
+        const documentSelection = document.getSelection();
+        const hasFocus = value.getSelection().getHasFocus();
+        const hasSelectedText = hasFocus && !documentSelection?.isCollapsed;
+        const selectionLinkEntity = getSelectionEntity(value, "LINK");
+        const selectionButtonEntity = getSelectionEntity(value, "BUTTON");
+        const isLinkOrButton = !!(selectionLinkEntity || selectionButtonEntity);
+        const shouldShowInlineToolbar =
+            !readonly && (hasSelectedText || !!savedSelection || (hasFocus && isLinkOrButton));
+
+        if (selectionLinkEntity) {
+            setEditorArea(RichTextEditorArea.LinkPreview);
+        } else if (selectionButtonEntity) {
+            setEditorArea(RichTextEditorArea.ButtonPreview);
+        }
+
+        if (shouldShowInlineToolbar) {
+            const rangeRect = documentSelection?.getRangeAt(0).getBoundingClientRect();
+
+            if (rangeRect && selectionRectRef.current) {
+                selectionRectRef.current.style.width = `${rangeRect.width}px`;
+                selectionRectRef.current.style.height = `${rangeRect.height}px`;
+                selectionRectRef.current.style.top = `${rangeRect.y + window.scrollY}px`;
+                selectionRectRef.current.style.right = `${rangeRect.x}px`;
+                selectionRectRef.current.style.bottom = `${rangeRect.y}px`;
+                selectionRectRef.current.style.left = `${rangeRect.x + window.scrollX}px`;
+            }
+        }
+
+        popperUpdate && (await popperUpdate());
+        setShowInlineToolbar(shouldShowInlineToolbar);
+
+        setEditorState(value);
+    };
+
+    const keyBindingFn = (event: KeyboardEvent): string | null => {
```

Comment:
we could use [useReducer](https://reactjs.org/docs/hooks-reference.html#usereducer) to cleanup the event dispatching